CIP-0095/README.md

```diff
@@ -0,0 +1,678 @@
+
+### Error Types
+
+This specification inherits all Error Types defined in CIP-30.
+
+// todo: add more + avoid collisions with CIP-62?
+
+#### Extended APIError
+
+Here we define extensions to
+[CIP-30's Error Types](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0030#error-types).
+
+```ts
+type enum APIErrorCode {
+ // CIP-30 error codes
+ InvalidArgumentError = -5
+ UnknownChoiceError = -6
+ UnknownKey = -7
+ InvalidKey = -8
+}
+
+interface APIError {
+  code: APIErrorCode,
+  info: string
+}
+```
+
+- `InvalidArgumentError` - Generic error for errors in the formatting of the
+  arguments.
+- `UnknownChoiceError` - client supplies an unknown choice in a `Vote`.
+- `UnknownKey` - client supplies an unknown public key (DRep or Stake) with the
+  expectation of signature.
+- `InvalidKey` - client supplies a public key that does not match another
+  supplied credential. This could be for DRep Keys not matching DRepID in
+  Retirement Certificate or, could be when the incorrect role is supplied in a
+  vote with a DRep key.
+
+##### APIError elements
+
+- `code`: The `APIErrorCode` which is being reported.
+- `info` - A human readable description of the error.
+
+#### Extended TxSignError
+
+Here we add additional error codes to CIP-30's
+[TxSignError](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0030#txsignerror).
+
+```ts
+type enum TxSignErrorCode {
+  ProofGeneration = 1,
+  UserDeclined = 2,
+}
+
+interface TxSignError = {
+  code: TxSignErrorCode,
+  info: string,
+}
+```
+
+- `ProofGeneration` - Raised when there is an error during witness/proof
+  generation.
+- `UserDeclined` - Raised when the user declined to sign one or many items.
+
+### Full Governance API
+
+Methods contained in this section on invocation should request the user to
+review and to consent to signature and submission. The exceptions to this are
+`.getActiveStakeKeys()` and `.getDRepKey()`, user consent should not be needed
+to share public key information.
+
+#### `api.getDRepKey(): Promise<PubDRepKey>`
+
+Errors: `APIError`
+
+The connected wallet account provides the account's public DRep Key, derivation
+as described in [DRep Key](#DRep-key).
+
+##### Returns
+
+The wallet account's DRep Key.
+
+#### `api.getActiveStakeKeys(): Promise<PubStakeKey[]>`
+
+Errors: `APIError`
+
+The connected wallet account's active public stake keys (with keys which are
+being used for staking), if the wallet tracks the keys that are used for
+governance then only those keys shall be returned.
+
+These are used by the client to identify the user's on-chain CIP-1694
+interactions. Here we allow for multiple stake credentials to provided at once
+for the case of
+[multi-stake key wallets](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0018).
+
+##### Returns
+
+An array of the connected user's active stake keys.
+
+#### `api.submitVoteDelegation([tx: UnsignedTransaction, stakeKey: PubStakeKey]): Promise<SubmittedTransaction[]>`
```

Comment:
❔ Would it be pertinent to allow voting with all your stake at onces?  using all stake keys?
Perhaps leaving `stakeKey` null could signal this intention and the wallet would include them all.

Reply:
This is an interesting idea, but I don't see the benefit of removing stakeKey from the call as it's inclusion gives the dApp an explicit way to be able to track the governance state of each key.

Reply:
- `stakeKey` arg might be redundant here. The Vote Delegation Certificate from the unsigned transaction already has a stakeKey, so it can be inferred (and should be inspected anyway to be validated by the user).
- If the above is true, then a single arg `submitVoteDelegation(tx: UnsignedTransaction)` would be sufficient, addressing voting delegation by adding multiple certificates in the same transaction

Reply:
Good point @mirceahasegan !